src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/RazorFormattingPass.cs

```diff
@@ -86,11 +86,172 @@ private static IEnumerable<TextEdit> FormatRazor(FormattingContext context, Razo
+                    closeBraceNode = child;
+                    closeBraceRange = child.GetRangeWithoutWhitespace(source);
+                }
+            }
+            // void Method()
+            // {
+            //     <div></div>
+            // }
+            else if (node is MarkupBlockSyntax markupBlockNode &&
+                markupBlockNode.Parent is CSharpCodeBlockSyntax cSharpCodeBlock)
+            {
+                for (var i = 0; i < cSharpCodeBlock.Children.Count; i++)
+                {
+                    var child = cSharpCodeBlock.Children[i];
+                    if (ReferenceEquals(child, markupBlockNode))
+                    {
+                        var previous = i == 0 ? 0 : i - 1;
+                        var next = i == cSharpCodeBlock.Children.Count - 1 ? i : i + 1;
+                        codeNode = markupBlockNode;
+                        codeRange = markupBlockNode.GetRangeWithoutWhitespace(source);
+                        var previousChild = cSharpCodeBlock.Children[previous];
+                        var nextChild = cSharpCodeBlock.Children[next];
+
+                        openBraceNode = previousChild;
+                        openBraceRange = previousChild.GetRangeWithoutWhitespace(source);
+                        closeBraceNode = nextChild;
+                        closeBraceRange = nextChild.GetRangeWithoutWhitespace(source);
+                    }
+                }
+            }
+            // We're looking for a code block like this:
+            //
+            // @{
+            //     var x = 1;
+            // }
+            else if (node is CSharpCodeBlockSyntax expliciteCode &&
+                expliciteCode.Children.First() is CSharpStatementSyntax statement &&
+                statement.Body is CSharpStatementBodySyntax csharpStatementBody)
+            {
+                openBraceNode = csharpStatementBody.OpenBrace;
+                openBraceRange = csharpStatementBody.OpenBrace.GetRange(source);
+                codeNode = csharpStatementBody.CSharpCode;
+                codeRange = csharpStatementBody.CSharpCode.GetRangeWithoutWhitespace(source);
+                closeBraceNode = csharpStatementBody.CloseBrace;
+                closeBraceRange = csharpStatementBody.CloseBrace.GetRange(source);
+            }
+            // @functions
+            // {
+            // }
+            else if (node is CSharpCodeBlockSyntax directiveCode &&
+                directiveCode.Children.Count == 1 && directiveCode.Children.First() is RazorDirectiveSyntax directive &&
+                directive.Body is RazorDirectiveBodySyntax directiveBody &&
+                directiveBody.Keyword.GetContent().Equals("functions"))
+            {
+                var cSharpCode = directiveBody.CSharpCode;
+                var openBrace = cSharpCode.Children.First(c => c.Kind == SyntaxKind.RazorMetaCode);
+                var closeBrace = cSharpCode.Children.Last(c => c.Kind == SyntaxKind.RazorMetaCode);
```

Comment:
Should these check the character too? I don't know what other things might be "meta code", or what if the code is malformed and there is only an open brace, is `closeBrace` going to point to that too?

Might be also worth using `FirstOrDefault` for all of these, and bailing if the result is null. I don't trust our test coverage enough to know that this won't blow up on a user.

Reply:
I don't know of any place we use MetaCode other than `{`, `}` and `@`, but it should be cheap to check.